### ISSUE TITLE
CUDA median filter behavior is broken. 

### ISSUE THREAD

**Issue by Mainvooid** [opencv/opencv_contrib#1799]
##### System information (version)

- OpenCV With Contrib => 3.4.3
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017
- CUDA =>9.2

```cpp
#include "opencv2/opencv.hpp"
#include "opencv2/cudafilters.hpp"
using namespace cv;
using namespace std;

int main() {
	string filename = "d_foreground.jpg";
	cv::Mat src= cv::imread(filename);
	cv::Mat img;
	cv::cvtColor(src, img, CV_BGRA2GRAY);

	/*----------------medianFilter -> GPU Test----------------------*/
	cv::cuda::Stream stream;
	cv::cuda::GpuMat d_img;
	d_img.upload(img, stream);
	cv::Ptr<cv::cuda::Filter> medianFilter = cv::cuda::createMedianFilter(CV_8UC1, 5);
	medianFilter->apply(d_img, d_img,stream);
	cv::Mat img_after_medianFilter;
	d_img.download(img_after_medianFilter, stream);
	stream.waitForCompletion();

	/*----------------medianBlur -> CPU Test----------------------*/
	cv::Mat img_after_medianBlur;
	cv::medianBlur(img, img_after_medianBlur, 5);

	/*-------------------------show-------------------------------*/
	cv::namedWindow("src", WINDOW_NORMAL);
	cv::namedWindow("img_after_medianFilter", WINDOW_NORMAL);
	cv::namedWindow("img_after_medianBlur", WINDOW_NORMAL);

	cv::resizeWindow("src",cv::Size(480,270));
	cv::resizeWindow("img_after_medianFilter", cv::Size(480, 270));
	cv::resizeWindow("img_after_medianBlur", cv::Size(480, 270));

	cv::imshow("src", src);
	cv::imshow("img_after_medianFilter", img_after_medianFilter);
	cv::imshow("img_after_medianBlur", img_after_medianBlur);
	cv:waitKey(0);
}

```
d_foreground.jpg
![d_foreground](https://user-images.githubusercontent.com/20666921/46239381-8f501380-c3ca-11e8-886a-9767fb6d6c25.jpg)

##### result
![img](https://user-images.githubusercontent.com/20666921/46239345-18b31600-c3ca-11e8-8db8-308e85005bd4.jpg)

The abnormal image is obtained. After binarization and median filtering  of CUDA , all zero values in the binary image will be changed to 128.

#### Solution
```cpp
cv::parallel_for_(Range(0, image.rows), [&](const Range& range)
{
	for (int rowIdx = range.start; rowIdx < range.end; rowIdx++)
	{
		uchar *pData = (uchar*)image.row(rowIdx).data;
		for (int i = 0; i < image.cols; i++)
			if (pData[i] == 128)
			{
				pData[i] = 0;
			}
	}
});
```

or

```cpp
for (int i = 0; i < image.rows; i++) {
	uchar *pData= image.ptr<uchar>(i);
	for (int j = 0; j <image.cols; j++) {
		if (pData[j] == 128) {
			pData[j] = 0;
		};
	}
}
```

[The same issue ](https://github.com/opencv/opencv/issues/10993)

**Comment by 1Reinier** [comment]
Did this get fixed in #12722?

**Comment by Mainvooid** [comment]
Repaired under `CUDA10+OPENCV4.0.1`
